File: RayTracerAvalonia/RayTracing/Shapes/Plane.cs
using System;
using System.Collections.Generic;
using System.Numerics;

namespace RayTracerAvalonia.RayTracing.Shapes;
public struct Plane : IShape
{
    private readonly Vector3 normal;
    private readonly float distance;
    public IMaterial Appearance { get; }

    public Plane(Vector3 normal, float distance, IMaterial texture)
    {
        Appearance = texture;
        this.normal = normal;
        this.distance = distance;
    }

    private readonly List<float> _emptyList = [];
    public readonly List<float> FindIntersections(Ray ray)
    {
        var angle = Vector3.Dot(ray.Direction, normal);

        // If the dot-product is zero, the ray is perpendicular to the plane's normal,
        // therefore the ray is parallel to the plane and will never intersect.
        if (Math.Abs(angle) < float.Epsilon)
            return _emptyList;

        var b = Vector3.Dot(normal, ray.Start + normal * -distance);
        return [-b / angle];
    }

    public readonly Vector3 GetNormalAt(Vector3 _) => normal;

    public readonly List<float> Intersect(Ray ray) => FindIntersections(ray);
}

File: RayTracerAvalonia.Desktop/cpp_raytracing_engine/plane.cpp
#include "include/plane.hpp"
#include <cmath>
#include <limits>

auto Plane::getColor() const -> Color { return color; }

auto Plane::intersect(const Ray &ray) const noexcept -> std::vector<float> {
  float angle = normal.dot(ray.direction);

  // If the dot-product is zero, the ray is parallel to the plane
  // and will never intersect
  if (std::abs(angle) < std::numeric_limits<float>::epsilon()) {
    return {};
  }

  float b = normal.dot(ray.start + normal * (-distance));
  float t = -b / angle;

  return {t};
}

auto Plane::getNormalAt(Vector3f point) const noexcept -> Vector3f {
  // For a plane, the normal is constant everywhere
  return normal;
}
